src/transaction.rs

```diff
@@ -31,7 +32,15 @@ where
 
     /// Commit the current transaction changes
     pub fn commit(mut self) -> Result<(), FbError> {
-        self.data.commit_retaining(self.conn)
+        let result = self.data.commit(self.conn);
+
+        if result.is_ok() {
+            ManuallyDrop::new(self);
+        } else {
+            let _ = self.rollback();
+        }
+
+        result
```

Comment:
If the commit fails, the transaction is not wrapped in `ManuallyDrop`, so the `rollback` is called twice.
```suggestion
        let result = self.data.commit(self.conn);

        let mut tr = ManuallyDrop::new(self);

        if result.is_err() {
            let _ = tr.data.rollback(tr.conn);
        }

        result
```
Edit: Fixed the suggestion

Reply:
Maybe it's better to just remove the explicit `rollback` and let drop do the job in case of failure

Reply:
In my change to `Transaction::commit`, the explicit rollback call on error is `Transaction::rollback` via `self.rollback()` (not `TransactionData::rollback`), so self is consumed by that `rollback()` call, and wrapped in `ManuallyDrop` there unconditionally.

Reply:
Oh, true, my mistake